design-proposal: instancetype.kubevirt.io - Expand during submission (#326)
kubevirt/community: design-proposals/instancetype.kubevirt.io/expand-during-submission.md

# Overview

The initial design of instance types and preferences captured a point in time copy of the original resource for future use by the VirtualMachine to ensure we always generate the same VirtualMachineInstance at runtime.

This additionally allows users to easily switch between classes, sizes and generations of these resources at some point in the future. However this flexibility comes at the cost of complexity within KubeVirt itself especially handling more advanced VirtualMachine lifecycle flows.

This design proposal aims to set out new configurable behavior where any referenced instance types and/or preferences are expanded into the VirtualMachine during submission.

## Motivation

TBD

## Goals

* Provide simple configurables at the cluster and VirtualMachine level to expand instance types and preferences during submission

## Non Goals

* The default behavior will not change as part of this work

## User Stories

* As a cluster owner I want to expand all new VirtualMachines referencing instance types and preferences during submission  

> * As a VirtualMachine owner I want to optionally expand my VirtualMachine referencing an instance type or preference

Reviewer: I think both user stories should state that it is an option that can be enabled?

> As a cluster owner I want to configure whether all new VirtualMachines referencing instance types and preferences are expanded during submission

> As a VirtualMachine owner I want to choose whether my VirtualMachine referencing an instance type or preference is expanded during submission

## API Examples

```go
// KubeVirtConfiguration holds all kubevirt configurations
type KubeVirtConfiguration struct {
[..]
	// Instancetype configuration
	Instancetype *InstancetypeConfiguration `json:"instancetype,omitempty"`

	// Preference configuration
	Preference *PreferenceConfiguration `json:"preference,omitempty"`
}

type InstancetypeConfiguration struct {
	Expand *bool `json:"expand,omitempty"`
}

type PreferenceConfiguration struct {
	Expand *bool `json:"expand,omitempty"`
}

[..]

// InstancetypeMatcher references a instancetype that is used to fill fields in the VMI template.
type InstancetypeMatcher struct {
    [..]
	Expand *bool `json:"expand,omitempty"`
}

[..]

// PreferenceMatcher references a set of preference that is used to fill fields in the VMI template.
type PreferenceMatcher struct {
    [..]
    Expand *bool `json:"expand,omitempty"`
}

```

## Scalability

The mutation and validation webhooks both already expand instance types and preferences into a local copy of the `VirtualMachine` during submission.

> The additional call to `Update` the `VirtualMachine` will introduce additional load across the control plane.

Author: Yeah given the size of the patch I don't think that would get us anything.

flexibility comes at the cost of complexity within KubeVirt itself when handling
more advanced `VirtualMachine` lifecycle flows.

This design proposal aims to set out new alternative and configurable behavior
where these revisions are no longer captured and are instead expanded into the
`VirtualMachine`.

## Motivation

The complexity of managing these point in time revisions of instance types and
preferences has grown over time as more features and versions of the CRDs have
landed.

Some of this complexity is exposed to users and third party integrations
such as back-up tooling or user-interfaces built on top of KubeVirt. This can
take the form of differing entry points for certain functionality such as hot
plug to requiring knowledge of the stored revisions to allow for a valid back up
and eventual restore of a given VirtualMachine to happen.

## Goals

* Provide a simple cluster configurable to control how
  instance types and preferences are referenced from a VirtualMachine

## Non Goals

* The default behavior will not change as part of this work

## User Stories

* As a cluster owner I want to control how instance types and preferences are

>   referenced from VirtualMachines within my environment

Author: The suggested configurable still allows a VirtualMachine to reference an instance type and preference so to me it doesn't make sense to focus on expansion here.

landed.

Some of this complexity is exposed to users and third party integrations
such as back-up tooling or user-interfaces built on top of KubeVirt. This can
take the form of differing entry points for certain functionality such as hot
plug to requiring knowledge of the stored revisions to allow for a valid back up
and eventual restore of a given `VirtualMachine` to happen.

## Goals

* Provide simple configurables at the cluster and `VirtualMachine` level to expand
  instance types and preferences

## Non Goals

* The default behavior will not change as part of this work

## User Stories

* As a cluster owner I want to configure the expansion of all new
  `VirtualMachines` referencing instance types and preferences
* As a VirtualMachine owner I want to optionally enable or disable the expansion
  my VirtualMachine referencing an instance type or preference

## Repos

* kubevirt/kubevirt

## Design

A new `KubeVirt` configurable will be introduced to have `VirtualMachines`
referencing instance types or preferences expanded.

Additional `VirtualMachine` configurables will also be introduced to allow a
specific newly submitted VirtualMachine referencing instance types or

> preferences to be expanded.

Reviewer 2: Can we consider alternatives where this is not available? My question stems form me not being aware of the second user story (a VM owner wanting to control this). I heard only of cluster admins.

Author: I can capture that in alternatives but personally I think that VM Owners should have the ability to expand or not regardless of the admins choice.

Reviewer 2: Thanks. But if I understand correctly, the biggest issue today is the complexity for third parties. If each VM owner can choose whatever they want, the complexity remains.

I would prefer have an opinionated solution, with less knobs and options; if we see a use case for the VM owners having this freedom to work against their cluster admin, we can add per-VM knob later. Removing knobs is practically impossible.

Author: That complexity will remain still even if this configurable is limited to cluster admins surely, especially given that we are not changing the default here.

Making this a cluster admin only configurable will mean that the end user experience of using a VirtualMachine initially defined with instance types and preferences can change drastically between environments and there's nothing a user can do to avoid this.

By providing an additional VM owner configurable they can be explicit and know that the resulting VirtualMachine will behave as expected.

Reviewer 2: > Users will also be unable to resize their `VirtualMachines` by making a singular
choice of instance type in the future

Why? A cluster admin that would like to interact with a third party can configure `expand` on her cluster, and that's that. If we complicate the solution and let the VM owner choose his own semantic, the cluster admin is left with uncertainty.

> Making this a cluster admin only configurable will mean that the end user experience of using a VirtualMachine initially defined with instance types and preferences can change drastically between environments and there's nothing a user can do to avoid this.

Correct. And that's exactly what I, as a cluster admin, want. I need to interact with a third party that is unaware of instance types, so I want my cluster to comply.

> By providing an additional VM owner configurable they can be explicit and know that the resulting VirtualMachine will behave as expected.

Right, but this adds complexity to the code, which I am not aware was requested by any user, increases development time, expands test matrix and add possibility of bugs, and cannot be dropped out of the API in the future.

Author: Okay these are fair points, I'll remove the user configurables for now and we can always follow up in the future if requested.

configurable in situations where they are not equal.

When requested the `VirtualMachine` controller will expand any referenced
instance type or preference when the `VirtualMachine` is reconciled.

Note that this will only occur when no revision has already been collected for
the resource and as such wouldn't impact existing `VirtualMachines` unless a
user explicitly requests to expand by removing the existing reference to the
revision.

## Concerns

### Exposing users to API complexity within VirtualMachines

One of the original design goals with instance types and preferences was to
simplify creation by reducing a users exposure to the core
`VirtualMachineInstanceSpec` API and thus their overall decision matrix.

While this proposal doesn't change the ability to simplify creation of
`VirtualMachines` it can result in a fully flattened `VirtualMachine` exposing
all of this complexity at that level once again.

### Breaking declarative management of VirtualMachines using instance types

This expansion behavior will break any declarative management of these
`VirtualMachines` as they will substantially change after initial submission.

> Users will need to explicitly request to not expand their `VirtualMachines`

Reviewer 2: ```suggestion
VM owners will need to explicitly request to not expand their `VirtualMachines`
```

referencing instance types or preferences to avoid this.

## Alternatives

### Eventual expansion of *all* VMs

A possible alternative to this design proposal is to have the VirtualMachine
controller eventually expand all VirtualMachines referencing instance types or
preferences regardless of a revision already being stored.

This behaviour could be controlled by the same cluster-wide and VirtualMachine
specific configurables as suggested by this proposal.

This could however result in drastic changes to existing VirtualMachine objects

> without user interaction.

Reviewer 2: Not sure I follow: a user interaction took place if the configurable was set to `expand`.

Author: I'll reword to explicitly state that this can happen without the VM owner interacting with the VM thanks.

We would also still need to retain webhook validation logic to ensure that a
given `VirtualMachine` using an instance type or preference is valid during

> submission.

Reviewer 2: I find it recommended for KubeVirt to move the validation logic from the webhook into the controller. Validating VM definition against cluster-level objects such is instance types or preferences is raceful by design and [breaks eventual consistency](https://docs.google.com/presentation/d/1EIuMwabaOcBw6AqcZCqz8Mip0HuhOcJtOsqVV0lVbEM/).

But I do not see how this is related to the design proposal. Can you elaborate?

Author: I understand but while there are still validations taking place in the webhooks we will need to expand any provided instance type and preference to retain the same behaviour as a flat VirtualMachine. Once these validations are moved into the control loop then yes the need for this is also removed.

I wanted to mention this in passing as it's logic we would still need to carry in the project that isn't removed by eventually expanding later.

Reviewer 2: does the following represents what you are saying?
```suggestion
Note that this proposal does not affect the validation logic to ensure that a
given `VirtualMachine` using an instance type or preference is valid during
submission.
```

> This would also put additional load on the controller if the cluster-wide

Reviewer 2: ```suggestion
This would also put one-off load on the controller if the cluster-wide
```
I think that on-startup load is reduced by approximately same amount. This option just makes it happen sooner.

Author: ACK yes fair point, changed.

kind: KubeVirt
metadata:
  name: kv
spec:
  configuration:
    instancetype:
      policy: expand
    preference:
      policy: expand
```

Additionally a VM owner can explicitly request `expand` by setting the policy
directly on the `VirtualMachine`:

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: cirros
spec:
  instancetype:
    name: foo
    policy: expand
  preference:
    name: bar
    policy: expand
```

Likewise users can explicitly request the `reference` policy to counter a
different default being defined in the `KubeVirt` `CR`:

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: cirros
spec:
  instancetype:
    name: foo
    policy: reference
```

Users can also expand an existing `VirtualMachine` by removing any referenced
revision from a matcher to invoke the expand behaviour, for example starting
with the following:

```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: cirros
spec:
  instancetype:
    name: foo
    revisionName: bar
```

A user can explicitly ask for the associated instance type to be expanded into
the `VirtualMachine` by removing `revisionName` and setting `policy` to

> `expand`:

Reviewer 2: Isn't adding `policy:expand` explicit enough? Why removing `revisionName` is required?

Author: The current proposal stops short of expanding all VirtualMachines and would use `revisionName` to determine if the VirtualMachine had already been seen by the controller and revisions stored. If we went with the alternative of expanding all regardless of a revision already being captured then `policy` being set to `expand` would be enough to trigger the behaviour.

    Policy *ExpansionPolicy `json:"policy,omitempty"`
}

[..]

// InstancetypeMatcher references a instancetype that is used to fill fields in the VMI template.
type InstancetypeMatcher struct {
    [..]
    Policy *ExpansionPolicy `json:"policy,omitempty"`
}

[..]

// PreferenceMatcher references a set of preference that is used to fill fields in the VMI template.
type PreferenceMatcher struct {
    [..]
    Policy *ExpansionPolicy `json:"policy,omitempty"`
}
```

## Scalability

The resulting mutation of the `VirtualMachine` with this proposal will
cause an additional reconciliation loop to trigger for the VM.

Work should be carried out to ensure that the substantial mutation of the
`VirtualMachine` during submission doesn't negatively impact the control plane.

## Update/Rollback Compatibility

Existing `VirtualMachines` referencing instance types or preferences will not be
changed as a result of this behaviour.

There will be no ability to automatically rollback new `VirtualMachines` once
they have their instance type or preference expanded by this new functionality.

Users will also be unable to resize their `VirtualMachines` by making a singular

> choice of instance type in the future.

Reviewer 2: Why not? It is still OK to define instance-typed VM `spec`s, why would it not be ok to modify an existing VM to be instance typed? In properly eventual-consistent systems, the intention of the user is what matters most, not the history of the resource.

Author: At present instance types conflict if the VirtualMachine already provides an amount of vCPU or memory.

We have talked about changing this behaviour in https://github.com/kubevirt/community/pull/318 and in doing so could allow a user to resize flat VirtualMachines by selecting an instance type but that might require a follow up design proposal and discussion to work out the details.

Reviewer 2: I do not follow... Assume that I created a flat VM with 7 CPUs. I stopped my VM and then changed my mind. I would like it to be much larger and more standard and edit the VM to use `cx1.8xlarge`. Would KubeVirt attempt to block this? Why?

Author: At present yes it would unless you also removed the 7 vCPUs you originally requested within the VirtualMachine.

We just don't allow you to request vCPUs and memory through the VirtualMachineInstanceSpec *and* instance type at the same time.

Reviewer 2: > At present yes it would unless you also removed the 7 vCPUs you originally requested within the VirtualMachine.
> 
> We just don't allow you to request vCPUs and memory through the VirtualMachineInstanceSpec _and_ instance type at the same time.

this is clear and expected. I just don't understand it from "Users will also be unable to resize their `VirtualMachines` by making a singular choice of instance type in the future". Would you agree to rephrase the text?

Author: Ah apologies I see your point, I'll rephrase now.
